File: 2012/s3/s3.py
# The Reading Class
class Reading:
	def __init__(self, r, f):
		self.reading = r
		self.frequency = f

#Initialize the frequencies list
F = []
for i in range(1001):
	F.append(Reading(i,0))

#Gets the number of sensors
N = input()

#Read input and increase frequency of read input
for i in range(N):
	F[input()].frequency += 1

#Sort the frequencies list 
for i in range(1, 1001):
	current = F[i]
	j = i - 1
	while j >= 0 and (F[j].frequency < current.frequency or (F[j].frequency == current.frequency and F[j].reading < current.reading)):
		F[j+1] = F[j]
		j = j-1
	F[j+1] = current

#Check for multiple same frequency readings
high = F[0].frequency == F[2].frequency
secondHigh = F[0].frequency != F[1].frequency and F[1].frequency == F[2].frequency

if high:
	i = 1
	while i < 1001 and F[0].frequency == F[i].frequency:
		i = i+1
	print abs(F[0].reading - F[i-1].reading)
elif secondHigh:
	i = 3
	m = abs(F[0].reading - F[1].reading)
	while i < 1001 and F[1].frequency == F[i].frequency:
		m = max(m, abs(F[0].reading - F[i].reading))
		i = i + 1 
	print m
else:
	print abs(F[0].reading - F[1].reading)
		

File: 2012/s3/s3.cpp
#include <iostream>
#include <cmath>

using namespace std;

class Reading
{
	public:
		int reading;
		int frequency;

		Reading() {reading = 0; frequency = 0;};
		Reading(int r, int f) {reading = r; frequency = f;};
};

int main()
{
	
	Reading readings[1000];

	for(int i = 0; i < 1000; ++i)
	{
		readings[i] = Reading(i+1,0);
	}

	int N = 0;
	cin >> N;

	for(int i = 0; i < N; ++i)
	{
		int reading = 0;
		cin >> reading;
		readings[reading-1].frequency++;
	}

	for(int i = 1; i < 1000; ++i)
	{
		Reading note = readings[i];
		int j = i-1;
		while(j >= 0 &&((readings[j].frequency <  note.frequency) || (readings[j].frequency == note.frequency && readings[j].reading < note.reading)))
		{
			readings[j+1] = readings[j];
			j--;
		}
		
		readings[j+1] = note;
	}

	if(readings[0].frequency == readings[1].frequency)
	{
		int i = 1;
		int max = 0;
		while(readings[i].frequency == readings[i-1].frequency)
		{
			max = readings[0].reading - readings[i].reading;
			i++;
		}

		cout << max << endl;

	}
	else if(readings[1].frequency != readings[2].frequency)
	{
		cout << abs(readings[0].reading - readings[1].reading) << endl;
	}
	else
	{
		int max = abs(readings[0].reading - readings[1].reading);
		int i = 2;
		while(readings[i-1].frequency == readings[i].frequency)
		{
			if(abs(readings[0].reading - readings[i].reading) > max)
			{
				max = abs(readings[0].reading - readings[i].reading);
			}
			i++;
		}

		cout << max << endl;
	}

	return 0;
}
